core/src/main/java/io/grpc/DnsNameResolver.java

```diff
@@ -85,39 +94,44 @@
   }
 
   @Override
-  public String getServiceAuthority() {
+  public final String getServiceAuthority() {
     return authority;
   }
 
   @Override
-  public synchronized void start(Listener listener) {
-    Preconditions.checkState(executor == null, "already started");
-    executor = SharedResourceHolder.get(GrpcUtil.SHARED_CHANNEL_EXECUTOR);
-    this.listener = listener;
+  public final synchronized void start(Listener listener) {
+    Preconditions.checkState(this.listener == null, "already started");
+    scheduledExecutor = SharedResourceHolder.get(scheduledExecutorResource);
+    this.listener = Preconditions.checkNotNull(listener, "listener");
     resolve();
   }
 
   @Override
-  public synchronized void refresh() {
-    Preconditions.checkState(executor != null, "not started");
+  public final synchronized void refresh() {
+    Preconditions.checkState(listener != null, "not started");
     resolve();
   }
 
-  @GuardedBy("this")
-  private void resolve() {
-    if (resolving) {
-      return;
-    }
-    resolving = true;
-    final Listener savedListener = Preconditions.checkNotNull(listener);
-    executor.execute(new Runnable() {
+  private final Runnable resolutionRunnable = new Runnable() {
       @Override
       public void run() {
         InetAddress[] inetAddrs;
+        Listener savedListener;
+        synchronized (DnsNameResolver.this) {
+          savedListener = listener;
+          resolving = true;
+          if (resolutionTask != null) {
+            resolutionTask.cancel(false);
```

Comment:
Isn't `resolutionTask` guaranteed to be the currently-executing Runnable? As in, it is guaranteed to never be cancelled at this point because it is running right now?